odoo/tools/populate.py

```diff
@@ -1,180 +1,357 @@
+
+    class Many2manyModelWrapper:
+        def __init__(self, field):
+            self._name = field.relation  # a m2m doesn't have a _name, so we use the tablename
+            self._table = field.relation
+            self._inherits = {}
+            self.env = env
+            self._rec_names_search = []
+            # if the field is inherited, the column attributes are defined on the base_field
+            column1 = field.column1 or field.base_field.column1
+            column2 = field.column2 or field.base_field.column2
+            # column1 refers to the model, while column2 refers to the comodel
+            self._fields = {
+                field.column1: Many2oneFieldWrapper(self, column1, field.model_name),
+                field.column2: Many2oneFieldWrapper(self, column2, field.comodel_name),
+            }
+
+        def __repr__(self):
+            return f"<Many2manyModelWrapper({self._name!r})>"
+
+        def __eq__(self, other):
+            return self._name == other._name
+
+        def __hash__(self):
+            return hash(self._name)
+
+    # check if the relation is an existing model
+    if model := next((model for model in env.registry.models.values() if model._table == field.relation), None):
+        return env[model._name]  # `model` is a MetaModel, re-fetch from env to get the actual ORM model
+    # the relation is a relational table, return a wrapped version
+    return Many2manyModelWrapper(field)
+
+
+def populate_models(env, models, factors, separator_code):
+    """
+    Create factors new records using existing records as templates.
+
+    If a dependency is found for a specific model, but it isn't specified by the user,
+    it will inherit the factor of the dependant model.
+    """
+
+    def has_records(_model):
+        query = SQL('SELECT EXISTS (SELECT 1 FROM %s)', SQL.identifier(_model._table))
+        return _model.env.execute_query(query)[0][0]
+
+    to_process = deque(models)
+    populated = defaultdict(int)  # {model: int(old_max_id)}
+    while to_process:
+        model = to_process.popleft()
+        if not has_records(model):  # if there are no records, there is nothing to populate
+            continue
+        # if the model has _inherits, the delegated models need to have been populated before the current one
+        missing_dependencies = [m for model_name in model._inherits if (m := env[model_name]) not in populated]
+        if missing_dependencies:
+            pending_dependencies = [dep for dep in missing_dependencies if dep not in to_process]
+            if pending_dependencies:
+                to_process.extendleft([model] + pending_dependencies)
+                factors.update(dict.fromkeys(pending_dependencies, factors[model]))
+                continue
+        # models on the other end of X2many relation should also be populated (ex: to avoid SO with no SOL)
+        for field in (f for f in model._fields.values() if f.store and f.copy):
```

Comment:
Simpler is better than complex
```suggestion
        for field in model._fields.values():
            if field.store and field.copy:            
```